twins/service.go

```diff
@@ -252,20 +254,31 @@ func (ts *twinsService) ListTwins(ctx context.Context, token string, offset uint
 	return ts.twins.RetrieveAll(ctx, res.GetValue(), offset, limit, name, metadata)
 }
 
-func (ts *twinsService) ListStates(ctx context.Context, token string, offset uint64, limit uint64, id string) (StatesPage, error) {
+func (ts *twinsService) ListStates(ctx context.Context, token string, offset uint64, limit uint64, twinID string) (StatesPage, error) {
 	_, err := ts.auth.Identify(ctx, &mainflux.Token{Value: token})
 	if err != nil {
 		return StatesPage{}, ErrUnauthorizedAccess
 	}
 
-	return ts.states.RetrieveAll(ctx, offset, limit, id)
+	return ts.states.RetrieveAll(ctx, offset, limit, twinID)
 }
 
 func (ts *twinsService) SaveStates(msg *messaging.Message) error {
-	ids, err := ts.twins.RetrieveByAttribute(context.TODO(), msg.Channel, msg.Subtopic)
+	var ids []string
+
+	ctx := context.TODO()
+	channel, subtopic := msg.Channel, msg.Subtopic
+	ids, err := ts.twinCache.IDs(ctx, channel, subtopic)
 	if err != nil {
 		return err
 	}
+	if (len(ids)) < 1 {
+		ids, err = ts.twins.RetrieveByAttribute(ctx, channel, subtopic)
+		if err != nil {
+			return err
+		}
+		ts.twinCache.SaveIDs(ctx, channel, subtopic, ids)
```

Comment:
If `len(ids) < 1`, you can do `return ts.twinCache.SaveIDs(ctx, channel, subtopic, ids)`, so that you don't ignore possible error here; there is no need to proceed to iteration over an empty array, as well.